src/driveformatthread.cpp

```diff
@@ -1,215 +1,259 @@
+        }
+#else
+        // No fallback available, report the original error
+        QString errorMessage;
+        switch (result.error()) {
+            case rpi_imager::FormatError::kFileOpenError:
+                errorMessage = tr("Error opening device for formatting");
+                break;
+            case rpi_imager::FormatError::kFileWriteError:
+                errorMessage = tr("Error writing to device during formatting");
+                break;
+            case rpi_imager::FormatError::kFileSeekError:
+                errorMessage = tr("Error seeking on device during formatting");
+                break;
+            case rpi_imager::FormatError::kInvalidParameters:
+                errorMessage = tr("Invalid parameters for formatting");
+                break;
+            case rpi_imager::FormatError::kInsufficientSpace:
+                errorMessage = tr("Insufficient space on device");
+                break;
+            default:
+                errorMessage = tr("Unknown formatting error");
+                break;
+        }
+        
+        qDebug() << "Cross-platform formatting failed:" << errorMessage;
+        emit error(errorMessage);
+#endif
+    }
+    else
+    {
+        qDebug() << "Cross-platform disk formatter succeeded";
+        emit success();
+    }
+#elif defined(Q_OS_DARWIN)
+    QProcess proc;
+    QStringList args;
+    args << "eraseDisk" << "FAT32" << "SDCARD" << "MBRFormat" << _device;
+    proc.start("diskutil", args);
+    proc.waitForFinished();
+
+    QByteArray output = proc.readAllStandardError();
+    qDebug() << args;
+    qDebug() << "diskutil output:" << output;
+
+    if (proc.exitCode())
+    {
+        emit error(tr("Error partitioning: %1").arg(QString(output)));
+    }
+    else
+    {
+        emit success();
+    }
+
+#elif defined(Q_OS_LINUX)
+
+    if (::access(_device.constData(), W_OK) != 0)
+    {
+        /* Not running as root, try to outsource formatting to udisks2 */
+
+#ifndef QT_NO_DBUS
+        UDisks2Api udisks2;
+        if (udisks2.formatDrive(_device))
+        {
+            emit success();
+        }
+        else
+        {
+#endif
+            emit error(tr("Error formatting (through udisks2)"));
+#ifndef QT_NO_DBUS
+        }
+#endif
+
+        return;
+    }
+
+    // Unmount the device before formatting
+    unmount_disk(_device);
+
+    qDebug() << "Formatting device" << _device << "with clean-room implementation";
+
+    // Get device size
+    std::uint64_t deviceSize = getDeviceSize(_device);
+    qDebug() << "Device size:" << deviceSize << "bytes";
+
+    // Use our clean-room disk formatter
+    rpi_imager::DiskFormatter formatter;
+    auto result = formatter.FormatDrive(_device.toStdString());
+
+    if (!result) {
+        QString errorMessage;
+        switch (result.error()) {
+            case rpi_imager::FormatError::kFileOpenError:
+                errorMessage = tr("Error opening device for formatting");
```

Comment:
These error-message strings appear in multiple places - perhaps they ought to be in one central place? (just to avoid the possibility of them drifting out of sync)